File: matlab/sinc3d.m
function wtrans=sinc3d(ifl,a1,a2,a3,klocs_d1,klocs_d2,klocs_d3,q,tol,mode)

if(nargin<1), test_sinc3d; return; end

%  
% wtrans(j) = sum sinc(a1(k)-klocs_d1(j)) * sinc(a2(k)-klocs_d2(j)) * sinc(a3(k)-klocs_d3(j)) * q(j)
%              k
%
% ifl = sinc convention
%   0: sinc(x) = sin(x)/x
%   1: sinc(x)=sin(pi*x)/(pi*x)
% a1 = (real) evaluation locations in dimension 1
% a2 = (real) evaluation locations in dimension 2
% a3 = (real) evaluation locations in dimension 3
% klocs_d1 = (real) sample locations in dimension 1
% klocs_d2 = (real) sample locations in dimension 2
% klocs_d3 = (real) sample locations in dimension 3
% q = sample strengths
% tol = requested precision

% ensure vector inputs are correct size
q=q(:).';
klocs_d1=klocs_d1(:);
klocs_d2=klocs_d2(:);
klocs_d3=klocs_d3(:);

newtol=max(tol,1e-16);

rkmaxx=max(bsxfun(@max,zeros(size(klocs_d1)),abs(klocs_d1)));
rkmaxx=max(rkmaxx,max(bsxfun(@max,zeros(size(a1)),abs(a1))));
rkmaxy=max(bsxfun(@max,zeros(size(klocs_d2)),abs(klocs_d2)));
rkmaxy=max(rkmaxy,max(bsxfun(@max,zeros(size(a2)),abs(a2))));
rkmaxz=max(bsxfun(@max,zeros(size(klocs_d3)),abs(klocs_d3)));
rkmaxz=max(rkmaxz,max(bsxfun(@max,zeros(size(a3)),abs(a3))));

if ifl==1
    rkmaxx=pi*rkmaxx;
    rkmaxy=pi*rkmaxy;
    rkmaxz=pi*rkmaxz;
    klocs_d1=pi*klocs_d1;
    klocs_d2=pi*klocs_d2;
    klocs_d3=pi*klocs_d3;
    a1=pi*a1;
    a2=pi*a2;
    a3=pi*a3;
end
rsamp=3; % increase to impose higher accuracy; will increase runtime
nx=ceil(rsamp*round(rkmaxx+3)); 
ny=ceil(rsamp*round(rkmaxy+3));
nz=ceil(rsamp*round(rkmaxz+3));

if isequal(mode,'legendre')
    [xx,wwx]=lgwt(nx,-1,1); 
    [yy,wwy]=lgwt(ny,-1,1);
    [zz,wwz]=lgwt(nz,-1,1);
    [c,d,e]=ndgrid(xx,yy,zz);
    allxx=c(:);
    allyy=d(:);
    allzz=e(:);
    [f,g,h]=ndgrid(wwx,wwy,wwz);
    allww=f(:).*g(:).*h(:);
    h_at_xxyyzz=finufft3d3(klocs_d1,klocs_d2,klocs_d3,q,-1,newtol,allxx,allyy,allzz);
    wtrans=(1/8)*finufft3d3(allxx,allyy,allzz,h_at_xxyyzz.*allww,1,newtol,a1,a2,a3);
else

a=-1; b=1;
%e=min(round(21+(-1*log10(newtol)-2)*25/14),60); % increase (up to 60) to impose higher accuracy; will increase runtime
e=25; % increase (up to 60) to impose higher accuracy
load('newconstants.mat'); 
constants=constantcell{e};

if mod(nx,2)~=0; nx=nx+1; end % ensure even so that 0 is a quadrature point
n=nx; h=(b-a)/n;
aind=e+1; bind=aind+n;
xx=a-(e*h):h:a+(n+e)*h; xx=xx(:); 
ww_x=zeros(length(xx),1);
ww_x(aind)=0.5; ww_x(bind)=0.5; 
ww_x(aind+1:bind-1)=1;
for k=1:e
    ww_x(aind-k) = ww_x(aind-k) - constants(k);
    ww_x(aind+k) = ww_x(aind+k) + constants(k);
    ww_x(bind-k) = ww_x(bind-k) + constants(k);
    ww_x(bind+k) = ww_x(bind+k) - constants(k);
end
ww_x=ww_x*h;

if mod(ny,2)~=0; ny=ny+1; end % ensure even so that 0 is a quadrature point
n=ny; h=(b-a)/n;
aind=e+1; bind=aind+n;
yy=a-(e*h):h:a+(n+e)*h; yy=yy(:);
ww_y=zeros(length(yy),1);
ww_y(aind)=0.5; ww_y(bind)=0.5; 
ww_y(aind+1:bind-1)=1;
for k=1:e
    ww_y(aind-k) = ww_y(aind-k) - constants(k);
    ww_y(aind+k) = ww_y(aind+k) + constants(k);
    ww_y(bind-k) = ww_y(bind-k) + constants(k);
    ww_y(bind+k) = ww_y(bind+k) - constants(k);
end
ww_y=ww_y*h;

if mod(nz,2)~=0; nz=nz+1; end % ensure even so that 0 is a quadrature point
n=nz; h=(b-a)/n;
aind=e+1; bind=aind+n;
zz=a-(e*h):h:a+(n+e)*h; zz=zz(:);
ww_z=zeros(length(zz),1);
ww_z(aind)=0.5; ww_z(bind)=0.5; 
ww_z(aind+1:bind-1)=1;
for k=1:e
    ww_z(aind-k) = ww_z(aind-k) - constants(k);
    ww_z(aind+k) = ww_z(aind+k) + constants(k);
    ww_z(bind-k) = ww_z(bind-k) + constants(k);
    ww_z(bind+k) = ww_z(bind+k) - constants(k);
end
ww_z=ww_z*h;

[c,d,e]=ndgrid(xx,yy,zz);
allxx=c(:);
allyy=d(:);
allzz=e(:);
[f,g,h]=ndgrid(ww_x,ww_y,ww_z);
allww=f(:).*g(:).*h(:);

% Transform sources and modes to proper form for finufft3d1 and finufft3d2
msx=length(xx); msy=length(yy); msz=length(zz);
actual_unif_spacex=xx(2)-xx(1); 
actual_unif_spacey=yy(2)-yy(1);
actual_unif_spacez=zz(2)-zz(1);
if max(abs(klocs_d1))*actual_unif_spacex>pi || max(abs(a1))*actual_unif_spacex>pi || max(abs(klocs_d2))*actual_unif_spacey>pi || max(abs(a2))*actual_unif_spacey>pi || max(abs(klocs_d3))*actual_unif_spacey>pi || max(abs(a3))*actual_unif_spacey>pi
    fprintf('Error: cannot use finufft type 1/2 here; outside [-pi,pi]\n');
end
Lx=xx(1); Ly=yy(1); Lz=zz(1);
if mod(msx,2)==0
    DU1x=-msx/2;
else
    DU1x=(-msx+1)/2;
end
if mod(msy,2)==0
    DU1y=-msy/2;
else
    DU1y=(-msy+1)/2;
end
if mod(msz,2)==0
    DU1z=-msz/2;
else
    DU1z=(-msz+1)/2;
end
translationx=(DU1x-(Lx/actual_unif_spacex));
translationy=(DU1y-(Ly/actual_unif_spacey));
translationz=(DU1z-(Lz/actual_unif_spacez));

h_at_xxyyzz=finufft3d1(klocs_d1*actual_unif_spacex,klocs_d2*actual_unif_spacey,klocs_d3*actual_unif_spacez,q.'.*exp(1i*((actual_unif_spacex*klocs_d1*translationx) + (actual_unif_spacey*klocs_d2*translationy) + (actual_unif_spacez*klocs_d3*translationz))),-1,newtol,msx,msy,msz);
h_at_xxyyzz=h_at_xxyyzz(:);

translationx=((-1*DU1x)*actual_unif_spacex)+Lx;
translationy=((-1*DU1y)*actual_unif_spacey)+Ly;
translationz=((-1*DU1z)*actual_unif_spacez)+Lz;
temp=finufft3d2(a1*actual_unif_spacex,a2*actual_unif_spacey,a3*actual_unif_spacez,1,newtol,reshape(h_at_xxyyzz.*allww,msx,msy,msz));
wtrans=(1/8)*exp(1i*(translationx*a1+translationy*a2+translationz*a3)).*temp;

% For real inputs, only return real values. Otherwise, return complex
if isreal(q)
    wtrans=real(wtrans);
end
wtrans=wtrans(:);

end

% For real inputs, only return real values. Otherwise, return complex
% wtrans
if isreal(q)
    wtrans=real(wtrans);
end
end

function test_sinc3d
n=10000; ifl=1;
numeval=100;
precisions=[1e-2 1e-3 1e-4 1e-5 1e-6 1e-7 1e-8 1e-9 1e-10 1e-11 1e-12 1e-13 1e-14 1e-15];
precisions=precisions(1:3:length(precisions));
klocs_d1=-10+(50*rand(n,1));
klocs_d2=-10+(20*rand(n,1));
klocs_d3=-10+(20*rand(n,1));
q=complex(rand(1,n)*30,rand(1,n)*30); 
a1=-10+(20*rand(n,1));
a2=-10+(20*rand(n,1));
a3=-10+(20*rand(n,1));
tic;correct=superslowsinc3d(ifl,a1(1:numeval),a2(1:numeval),a3(1:numeval),klocs_d1,klocs_d2,klocs_d3,q);t3=toc;
for p=1:length(precisions)
    pr=precisions(p);
    tic;myresult=sinc3d(ifl,a1,a2,a3,klocs_d1,klocs_d2,klocs_d3,q,pr,'legendre');t1=toc;
    tic;myresult2=sinc3d(ifl,a1,a2,a3,klocs_d1,klocs_d2,klocs_d3,q,pr,'trap');t2=toc;
    err1=norm(correct-myresult(1:numeval),2)/norm(correct,2);
    err2=norm(correct-myresult2(1:numeval),2)/norm(correct,2);
    fprintf("Requested: %g Error (Leg): %g (Trap): %g\n", pr, err1,err2);
    fprintf("              Time  (Leg): %g s (Trap): %g s (Direct): %g s\n",t1,t2,t3);
end
end

function correct=slowsinc3d(ifl,a1,a2,a3,klocs_d1,klocs_d2,klocs_d3,q)
    q=q(:).';
    [a1,b1]=ndgrid(a1,klocs_d1);
    [a2,b2]=ndgrid(a2,klocs_d2);
    [a3,b3]=ndgrid(a3,klocs_d3);
    if ifl==1
        x=sin(pi*(a1-b1))./(pi*(a1-b1));
        y=sin(pi*(a2-b2))./(pi*(a2-b2));
        z=sin(pi*(a3-b3))./(pi*(a3-b3));
    else
        x=sin(a1-b1)./(a1-b1);
        y=sin(a2-b2)./(a2-b2);
        z=sin(a3-b3)./(a3-b3);
    end
    x(arrayfun(@isnan,x))=1;
    y(arrayfun(@isnan,y))=1;
    z(arrayfun(@isnan,z))=1;
    sincmat=x.*y.*z;
    correct=sum(repmat(q,length(klocs_d1),1).*sincmat,2);
end

function correct=superslowsinc3d(ifl,a1,a2,a3,klocs_d1,klocs_d2,klocs_d3,q) % wrong here??
% Alternative brute force calculation; even slower
% May be substituted in timing tests     
    results=zeros(size(a1)); 
    for ind=1:length(results)
        sm=0;
        for j=1:length(klocs_d1)
            if ifl==0
                val=q(j)*(sinc(a1(ind)-klocs_d1(j)))*(sinc(a2(ind)-klocs_d2(j)))*(sinc(a3(ind)-klocs_d3(j)));
            else
                val=q(j)*(sinc(pi*(a1(ind)-klocs_d1(j))))*(sinc(pi*(a2(ind)-klocs_d2(j))))*(sinc(pi*(a3(ind)-klocs_d3(j))));
            end
            sm=sm+val;
        end
        results(ind)=sm;
    end
    correct=results;
end

function val=sinc(x)
    if x==0
        val=1;
    else
        val=sin(x)/x;
    end
end

File: src/sinc3d.cpp
#include <iostream>
#include <iomanip>
#include <complex>
#include <math.h> 
#include <ctime> 
#include "finufft.h" // for non-uniform Fourier transform
#include "fastgl.hpp" //for Legendre-Gauss weights
#include "sincutil.hpp"
#include "sinctransform.hpp"

using namespace std;

// Contains sinc3d and sincsq3d

int sinc3d(int ifl,int numlocs,double *klocs_d1_,double *klocs_d2_,double *klocs_d3_,complex<double> *q,double tol,complex<double> *res)
{
	/*  
	Computes res[j] = sum sinc(klocs_d1_[k]-klocs_d1_[j]) * sinc(klocs_d2_[k]-klocs_d2_[j]) * sinc(klocs_d3_[k]-klocs_d3_[j]) * q[j]
	             	   k

	Input:
		ifl = sinc convention
			0: sinc(x) = sin(x)/x
			1: sinc(x)=sin(pi*x)/(pi*x)
		klocs_ = (real) sample locations
		q = sample strengths
		tol = requested precision

	Returns:
		0: if success
		Other: error code, as returned by finufft (see finufft documentation)
	*/
	double pi=4*atan(1); 
	double lowesttol=1e-15; // best precision that finufft can take?
	double newtol=(tol/1000 > lowesttol) ? tol/1000 : lowesttol;
	float rkmaxx=0;
	float rkmaxy=0;
	float rkmaxz=0;
	double* klocs_d1=(double*)malloc(sizeof(double)*numlocs);
	double* klocs_d2=(double*)malloc(sizeof(double)*numlocs);
	double* klocs_d3=(double*)malloc(sizeof(double)*numlocs);
	complex<double>* qc=(complex<double>*)malloc(sizeof(complex<double>)*numlocs);
	for(int a=0;a<numlocs;a++)
	{
		double a_d1=abs(klocs_d1_[a]); 
		double a_d2=abs(klocs_d2_[a]); 
		double a_d3=abs(klocs_d3_[a]); 
		if(a_d1>rkmaxx)
		{
			rkmaxx=a_d1;
		}
		if(a_d2>rkmaxy)
		{
			rkmaxy=a_d2;
		}
		if(a_d3>rkmaxz)
		{
			rkmaxz=a_d3;
		}
		qc[a]=q[a]; //make q complex for finufft1d3
		klocs_d1[a]=klocs_d1_[a]; // to ensure that program does not alter input
		klocs_d2[a]=klocs_d2_[a];
		klocs_d3[a]=klocs_d3_[a];
	}
	if (ifl==1)
	{
		rkmaxx=pi*rkmaxx;
		rkmaxy=pi*rkmaxy;
		rkmaxz=pi*rkmaxz;
		for(int a=0;a<numlocs;a++)
		{
			klocs_d1[a]=klocs_d1[a]*pi;
			klocs_d2[a]=klocs_d2[a]*pi;
			klocs_d3[a]=klocs_d3[a]*pi;
		}
	}
	int rsamp=2;
	int nx=ceil(rsamp*round(rkmaxx+3));
	int ny=ceil(rsamp*round(rkmaxy+3));
	int nz=ceil(rsamp*round(rkmaxz+3));

	//create grid
	double* allxx=(double*)malloc(sizeof(double)*nx*ny*nz);
	double* allyy=(double*)malloc(sizeof(double)*nx*ny*nz);
	double* allzz=(double*)malloc(sizeof(double)*nx*ny*nz);
	double* allww=(double*)malloc(sizeof(double)*nx*ny*nz);
	double tempx,tempwx,tempy,tempwy,tempz,tempwz;
	int counter=0;
	for(int a=0;a<ny;a++)
	{
		fastgl::QuadPair py = fastgl::GLPair((size_t)ny,(size_t)(a+1));
		tempy=py.x();
		tempwy=py.weight;
		for(int b=0;b<nx;b++)
		{
			fastgl::QuadPair px = fastgl::GLPair((size_t)nx,(size_t)(b+1));
			tempx=px.x();
			tempwx=px.weight;
			for(int c=0;c<nz;c++)
			{
				fastgl::QuadPair pz = fastgl::GLPair((size_t)nz,(size_t)(c+1));
				tempz=pz.x();
				tempwz=pz.weight;

				allxx[counter]=tempx;
				allyy[counter]=tempy;
				allzz[counter]=tempz;
				allww[counter]=tempwx*tempwy*tempwz;
				counter++;
			}
		}
	}
	nufft_opts* opts = new nufft_opts;
        finufft_default_opts(opts);
	//h_at_xx will be complex
	complex<double>* h_at_xxyyzz=(complex<double>*)malloc(sizeof(complex<double>)*nx*ny*nz);
	int ier1=finufft3d3(numlocs,klocs_d1,klocs_d2,klocs_d3,qc,-1,newtol,nx*ny*nz,allxx,allyy,allzz,h_at_xxyyzz,opts);
	if (ier1 != 0)
	{
		cout<<"Error: call 1 to finufft2d3 failed: "<<ier1<<"\n";
		return ier1;
	}
	complex<double>* weighted=(complex<double>*)malloc(sizeof(complex<double>)*nx*ny*nz);
	for(int a=0;a<(nx*ny*nz);a++)
	{
		weighted[a]=(1.0/8)*h_at_xxyyzz[a]*allww[a];
	}

	complex<double>* wtrans=(complex<double>*)malloc(sizeof(complex<double>)*numlocs);
	int ier2=finufft3d3(nx*ny*nz,allxx,allyy,allzz,weighted,1,newtol,numlocs,klocs_d1,klocs_d2,klocs_d3,wtrans,opts);
	if (ier2 != 0)
	{
		cout<<"Error: call 2 to finufft2d3 failed: "<<ier2<<"\n";
		return ier2;
	}

	//make wtrans real (double array)
	for(int a=0;a<numlocs;a++)
	{
		res[a]=wtrans[a];
	}

	//FREE MEMORY
	free(wtrans);
	free(weighted);
	free(h_at_xxyyzz);
	free(qc);
	free(allxx);
	free(allyy);
	free(allzz);
	free(allww);
	return 0;
}


int sincsq3d(int ifl,int numlocs,double *klocs_d1_,double *klocs_d2_,double *klocs_d3_,complex<double> *q,double tol,complex<double> *res)
{
	/*  
	Computes res[j] = sum sinc^2(klocs_d1_[k]-klocs_d1_[j]) * sinc^2(klocs_d2_[k]-klocs_d2_[j]) * sinc^2(klocs_d3_[k]-klocs_d3_[j]) * q[j]
	             	   k

	Inputs:
		ifl = sinc convention
			0: sinc(x) = sin(x)/x
			1: sinc(x)=sin(pi*x)/(pi*x)
		klocs_ = (real) sample locations
		q = sample strengths
		tol = requested precision

	Returns:
		0: if success
		Other: error code, as returned by finufft (see finufft documentation)
	*/
	double pi=4*atan(1); 
	double lowesttol=1e-15; // best precision that finufft can take?
	double newtol=(tol/1000 > lowesttol) ? tol/1000 : lowesttol;
	float rkmaxx=0;
	float rkmaxy=0;
	float rkmaxz=0;
	double* klocs_d1=(double*)malloc(sizeof(double)*numlocs);
	double* klocs_d2=(double*)malloc(sizeof(double)*numlocs);
	double* klocs_d3=(double*)malloc(sizeof(double)*numlocs);
	complex<double>* qc=(complex<double>*)malloc(sizeof(complex<double>)*numlocs);
	for(int a=0;a<numlocs;a++)
	{
		double a_d1=abs(klocs_d1_[a]); 
		double a_d2=abs(klocs_d2_[a]); 
		double a_d3=abs(klocs_d3_[a]); 
		if(a_d1>rkmaxx)
		{
			rkmaxx=a_d1;
		}
		if(a_d2>rkmaxy)
		{
			rkmaxy=a_d2;
		}
		if(a_d3>rkmaxz)
		{
			rkmaxz=a_d3;
		}
		qc[a]=q[a]; //make q complex for finufft1d3
		klocs_d1[a]=klocs_d1_[a]; // to ensure that program does not alter input
		klocs_d2[a]=klocs_d2_[a];
		klocs_d3[a]=klocs_d3_[a];
	}
	if (ifl==1)
	{
		rkmaxx=pi*rkmaxx;
		rkmaxy=pi*rkmaxy;
		rkmaxz=pi*rkmaxz;
		for(int a=0;a<numlocs;a++)
		{
			klocs_d1[a]=klocs_d1[a]*pi;
			klocs_d2[a]=klocs_d2[a]*pi;
			klocs_d3[a]=klocs_d3[a]*pi;
		}
	}
	int rsamp=2;
	int nx=ceil(rsamp*round(rkmaxx+3));
	int ny=ceil(rsamp*round(rkmaxy+3));
	int nz=ceil(rsamp*round(rkmaxz+3));

	double *xx=(double*)malloc(sizeof(double)*nx*2);
	double *wwx=(double*)malloc(sizeof(double)*nx*2);
	double tempx,tempw;
	for(int a=1;a<=nx;a++)
	{
		fastgl::QuadPair p = fastgl::GLPair((size_t)nx,(size_t)a);
		tempx=p.x();
		tempw=p.weight;
		xx[a-1]=tempx-1;
		wwx[a-1]=tempw*(2-abs(xx[a-1]));
		xx[a+nx-1]=tempx+1;
		wwx[a+nx-1]=tempw*abs(2-abs(xx[a+nx-1]));

	}

	double *yy=(double*)malloc(sizeof(double)*ny*2);
	double *wwy=(double*)malloc(sizeof(double)*ny*2);
	for(int a=1;a<=ny;a++)
	{
		fastgl::QuadPair p = fastgl::GLPair((size_t)ny,(size_t)a);
		tempx=p.x();
		tempw=p.weight;
		yy[a-1]=tempx-1;
		wwy[a-1]=tempw*(2-abs(yy[a-1]));
		yy[a+ny-1]=tempx+1;
		wwy[a+ny-1]=tempw*(2-abs(yy[a+ny-1]));
	}

	double *zz=(double*)malloc(sizeof(double)*nz*2);
	double *wwz=(double*)malloc(sizeof(double)*nz*2);
	for(int a=1;a<=nz;a++)
	{
		fastgl::QuadPair p = fastgl::GLPair((size_t)nz,(size_t)a);
		tempx=p.x();
		tempw=p.weight;
		zz[a-1]=tempx-1;
		wwz[a-1]=tempw*(2-abs(zz[a-1]));
		zz[a+nz-1]=tempx+1;
		wwz[a+nz-1]=tempw*(2-abs(zz[a+nz-1]));
	}

	//create grid
	double* allxx=(double*)malloc(sizeof(double)*8*nx*ny*nz);
	double* allyy=(double*)malloc(sizeof(double)*8*nx*ny*nz);
	double* allzz=(double*)malloc(sizeof(double)*8*nx*ny*nz);
	double* allww=(double*)malloc(sizeof(double)*8*nx*ny*nz);
	int counter=0;
	for(int a=0;a<(2*nz);a++)
	{
		for(int b=0;b<(2*ny);b++)
		{
			for(int c=0;c<(2*nx);c++)
			{
				allxx[counter]=xx[c];
				allyy[counter]=yy[b];
				allzz[counter]=zz[a];
				allww[counter]=wwx[c]*wwy[b]*wwz[a];
				counter++;
			}
		}
	}

	nufft_opts* opts = new nufft_opts;
        finufft_default_opts(opts);
	//h_at_xx will be complex
	complex<double>* h_at_xxyyzz=(complex<double>*)malloc(sizeof(complex<double>)*8*nx*ny*nz);
	int ier1=finufft3d3(numlocs,klocs_d1,klocs_d2,klocs_d3,qc,-1,newtol,8*nx*ny*nz,allxx,allyy,allzz,h_at_xxyyzz,opts);
	if (ier1 != 0)
	{
		cout<<"Error: call 1 to finufft2d3 failed: "<<ier1<<"\n";
		return ier1;
	}
	complex<double>* weighted=(complex<double>*)malloc(sizeof(complex<double>)*8*nx*ny*nz);
	for(int a=0;a<(8*nx*ny*nz);a++)
	{
		weighted[a]=(1.0/64)*h_at_xxyyzz[a]*allww[a];
	}

	complex<double>* wtrans=(complex<double>*)malloc(sizeof(complex<double>)*numlocs);
	int ier2=finufft3d3(8*nx*ny*nz,allxx,allyy,allzz,weighted,1,newtol,numlocs,klocs_d1,klocs_d2,klocs_d3,wtrans,opts);
	if (ier2 != 0)
	{
		cout<<"Error: call 2 to finufft2d3 failed: "<<ier2<<"\n";
		return ier2;
	}

	for(int a=0;a<numlocs;a++)
	{
		res[a]=wtrans[a];
	}
	free(wtrans);
	free(weighted);
	free(h_at_xxyyzz);
	free(qc);
	free(allxx);
	free(allyy);
	free(allww);
	free(xx);
	free(yy);
	free(zz);
	free(wwx);
	free(wwy);
	free(wwz);
	return 0;
}

